packages/server/lib/cloud/environment.ts

```diff
@@ -0,0 +1,165 @@
+  const processTree = stdout.split('\n').reduce((acc, line) => {
+    const [pid, ppid] = line.trim().split(/\s+/)
+
+    acc.set(pid, ppid)
+
+    return acc
+  }, new Map())
+
+  const currentProcessBranch: string[] = []
+
+  while (pid && pid !== '0') {
+    currentProcessBranch.push(pid)
+    pid = processTree.get(pid)
+  }
+
+  return currentProcessBranch
+}
+
+interface GetCypressEnvUrlFromProcessBranch {
+  envUrl?: string
+  error?: {
+    dependency?: string
+    name: string
+    message: string
+    stack: string
+  }
+}
+
+// See https://whimsical.com/encryption-logic-BtJJkN7TxacK8kaHDgH1zM for more information on what this is doing
+const getCypressEnvUrlFromProcessBranch = async (pid: string): Promise<GetCypressEnvUrlFromProcessBranch> => {
+  let error: { name: string, message: string, stack: string } | undefined
+  let envUrl: string | undefined
+
+  if (process.platform !== 'win32') {
+    try {
+      const processBranch = await getProcessBranchForPid(pid)
+      const { stdout } = await execAsync(`ps eww -p ${processBranch.join(',')} -o pid=,command=`)
+
+      const pidEnvUrlMapping = stdout.split('\n').reduce((acc, line) => {
+        const cypressEnvUrl = line.trim().match(/(\d+)\s.*CYPRESS_API_URL=(\S+)\s/)
+
+        if (cypressEnvUrl) {
+          acc.set(cypressEnvUrl[1], cypressEnvUrl[2])
+        }
+
+        return acc
+      }, new Map())
+
+      const foundPid = processBranch.find((pid) => pidEnvUrlMapping.get(pid))
+
+      if (foundPid) {
+        envUrl = pidEnvUrlMapping.get(foundPid)
+      }
+    } catch (err) {
+      error = err
+    }
+  }
+
+  return {
+    envUrl,
+    error,
+  }
+}
+
+interface DependencyInformation {
+  maybeCheckProcessTreeIfPresent: string[]
+  neverCheckProcessTreeIfPresent: string[]
+}
+
+// See https://whimsical.com/encryption-logic-BtJJkN7TxacK8kaHDgH1zM for more information on what this is doing
+const getEnvInformationForProjectRoot = async (projectRoot: string, pid: string) => {
+  let dependencies = {}
+  let errors: { dependency?: string, name: string, message: string, stack: string }[] = []
+  let envDependenciesVar = process.env.CYPRESS_ENV_DEPENDENCIES
+  let envUrl = process.env.CYPRESS_API_URL
+  let checkProcessTree
+
+  if (envDependenciesVar) {
+    const envDependenciesInformation = JSON.parse(base64Url.decode(envDependenciesVar)) as DependencyInformation
+
+    const packageToJsonMapping: Record<string, string> = {}
+
+    envDependenciesInformation.maybeCheckProcessTreeIfPresent.forEach((dependency) => {
+      try {
+        const packageJsonPath = resolvePackagePath(dependency, projectRoot)
+
+        if (packageJsonPath) {
+          packageToJsonMapping[dependency] = packageJsonPath
+          checkProcessTree = true
+        }
+      } catch (error) {
+        errors.push({
+          dependency,
+          name: error.name,
+          message: error.message,
+          stack: error.stack,
+        })
```

Comment:
might want to refactor this into a common function since it does the same thing as lines 121-126 below

Reply:
Updated here with cleaner logic: [`802f193` (#25837)](https://github.com/cypress-io/cypress/pull/25837/commits/802f19373fc7af205b04406440a08ba34f1e6a27)